cmd/zoekt-sourcegraph-indexserver/main.go

```diff
@@ -322,6 +330,59 @@ func (s *Server) Run(queue *Queue) {
 	}
 }
 
+// Update priority.json given new entries, and remove no longer tracked repos.
+// This doesn't simply write newPriorities because a transient getIndexOptions failure
+// would cause the associated repo to get deprioritized.
+func (s *Server) maybeUpdatePriorities(names []string, newPriorities map[string]float64) {
+	priorityPath := path.Join(s.IndexDir, "priority.json")
+	priorities := map[string]float64{}
+	buf, err := ioutil.ReadFile(priorityPath)
+	if err == nil {
+		err = json.Unmarshal(buf, &priorities)
+		if err != nil {
+			log.Printf("warning: error loading old priority.json, treating as empty: %v", err)
+		}
+	}
+
+	// maybe remove no-longer-tracked repos from the priorities list
+	if len(names) != len(priorities) {
+		set := make(map[string]struct{}, len(names))
+		for _, name := range names {
+			set[name] = struct{}{}
+		}
+		for name := range priorities {
+			if _, ok := set[name]; !ok {
+				delete(priorities, name)
+			}
+		}
+	}
+
+	for name, priority := range newPriorities {
+		priorities[name] = priority
+	}
+
+	newBuf, err := json.Marshal(priorities)
+	if err != nil {
+		log.Printf("error marshaling new priority.json: %v", err)
+	}
+	newBuf = append(newBuf, '\n') // prettier
+
+	if bytes.Equal(buf, newBuf) {
```

Comment:
I'm not sure if this is safe? Does `encoding/json` have a well-defined order it marshals keys? I would suspect it is safer to compare `newPriorities` and `priorities`. This is possible with my suggestion above where you only mutate `newPriorities`

Reply:
Yes, this is safe: https://golang.org/pkg/encoding/json/#Marshal

> Map values encode as JSON objects. The map's key type must either be a string, an integer type, or implement encoding.TextMarshaler. The map keys are sorted and used as JSON object keys by applying the following rules, subject to the UTF-8 coercion described for string values above: